Wip/5386

nirbar opened the pull request:
Support MSI transaction
Opt-in by specifying Transaction='yes' on RollbackBoundary

nirbar committed:
Initial- support rollback boundaries with MSI transactions.

nirbar committed:
- Handle per user/machine transaction contexts
- Disable transaction if no-rollback is on
- Fail installation if reboot requested during transaction.

nirbar committed:
Fix missing reference

nirbar committed:
Add transaction tests- initial

dnfclas commented:
Hi **@nirbar**, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -745,6 +771,36 @@ extern "C" HRESULT ApplyExecute(
+				else
+				{
+					LogString(REPORT_STANDARD, "Starting a new MSI transaction\n");
+					hr = DoMsiBeginTransaction(&context, pEngineState);
```
If we're really going to bail if the transaction can't be started and not fallback to burn's traditional rollback logic, then we should add new plan actions for starting/completing/rolling back the transaction instead of doing it in the BURN_EXECUTE_ACTION_TYPE_ROLLBACK_BOUNDARY action, and not even schedule rollback actions for each individual package.

rseanhall reviewed (commented):
I haven't done a thorough review but here's what I see.

You have comments that it only supports per-machine, but have a commit that looks like you tried to add support for per-user as well.  Is per-user supposed to work?

I like that you have a test, but would really want to see one where it's testing the rollback.

Also, we use spaces and it looks like there are tabs in the C++ code.

Commit:
Thicken transaction rollback test:
Rollback should rollback a package that performed a major upgrade.
The preceding package (the one that was upgraded from) should be present after the transaction rolled back.

Commit:
Add a transaction test that fails and rollsback the transaction

Commit:
Spaces for tabs

Commit:
During build, check that a MSI transaction has all x64 before any x86 package.

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -745,6 +771,36 @@ extern "C" HRESULT ApplyExecute(
+                    fInTransaction = TRUE;
+                }
+            }
+        }
```
We should not be performing any of these actions based on the rollback boundary action.  During the plan phase, we need to insert new actions, for example BURN_EXECUTE_ACTION_TYPE_BEGIN_MSI_TRANSACTION or BURN_EXECUTE_ACTION_TYPE_ROLLBACK_MSI_TRANSACTION.  In this way, the plan will have baked in whether a transaction is currently happening or not.  Most of the code that you've added here should then be moved to DoExecuteAction or DoRollbackActions.

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -763,7 +819,15 @@ extern "C" HRESULT ApplyExecute(
+            if (fInTransaction)
+            {
+                hr = E_INVALIDSTATE;
+                LogString(REPORT_ERROR, "Ilegal state: Reboot requested within an MSI transaction. Transaction will rollback.");
```
*Illegal

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -1628,6 +1701,127 @@ static void DoRollbackCache(
+}
+
+// Currently, supporting only elevated transactions.
+static HRESULT DoMsiBeginTransaction(
```
I would expect these DoMsi* methods to be deleted when the code is moved into DoExecuteAction/DoRollbackActions.

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -1628,6 +1701,127 @@ static void DoRollbackCache(
+        MSIHANDLE hMsiTrns = NULL;
+        HANDLE hMsiTrnsEvent = NULL;
+        uResult = MsiBeginTransaction(L"WiX", 0, &hMsiTrns, &hMsiTrnsEvent);
+        ExitOnWin32Error(uResult, hr, "Failed beginning an MSI transaction");
```
In each of these ExecuteMsi* methods, the actual implementation should live in an external method so that both the elevated code and unelevated code (eventually) call the same function.

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -1628,6 +1701,127 @@ static void DoRollbackCache(
+    {
+        MSIHANDLE hMsiTrns = NULL;
+        HANDLE hMsiTrnsEvent = NULL;
+        uResult = MsiBeginTransaction(L"WiX", 0, &hMsiTrns, &hMsiTrnsEvent);
```
[WiX code style #13](http://wixtoolset.org/development/code-style/) - Functions that are from a public SDK should be prefixed with the global namespace scope operator `::`.

rseanhall commented on src/burn/engine/apply.cpp:
```diff
@@ -1805,18 +2007,33 @@ static HRESULT DoRollbackActions(
+                {
+                    LogString(REPORT_STANDARD, "Skipping rolling back an MSI package- already done in transaction rollback\n");
+                    break;
+                }
```
After modifying the plan code to insert actions for MSI transactions, rollback actions for these packages should not be created in the plan during an MSI transaction.  So these checks should be asserts or deleted.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -34,6 +34,11 @@ typedef enum _BURN_ELEVATION_MESSAGE_TYPE
     BURN_ELEVATION_MESSAGE_TYPE_EXECUTE_FILES_IN_USE,
     BURN_ELEVATION_MESSAGE_TYPE_LAUNCH_APPROVED_EXE_PROCESSID,
+
+    BURN_ELEVATION_TRANSACTION_BEGIN,
```
These names should all have the prefix BURN_ELEVATION_MESSAGE_TYPE.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -69,11 +74,24 @@ typedef struct _BURN_ELEVATION_CHILD_MESSAGE_CONTEXT
     BURN_REGISTRATION* pRegistration;
     BURN_USER_EXPERIENCE* pUserExperience;
+
+	MSIHANDLE hMsiTrns;
```
This struct is for providing context for elevated actions, not for storing new data.  A new one is created for each incoming message and thrown away after processing it.  In addition, I didn't see us using this data anyway so let's remove them.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -69,11 +74,24 @@ typedef struct _BURN_ELEVATION_CHILD_MESSAGE_CONTEXT
 // internal function declarations
 
+static HRESULT OnMsiBeginTransaction(
+    __in BURN_ELEVATION_CHILD_MESSAGE_CONTEXT* pContext
```
All of the other On* methods in here specify exactly what they need from the BURN_ELEVATION_CHILD_MESSAGE_CONTEXT struct, so I would prefer that these methods don't take this as a parameter.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -701,6 +719,71 @@ extern "C" HRESULT ElevationExecuteExePackage(
+    __in LPVOID pvContext
+)
+{
+    UNREFERENCED_PARAMETER(hwndParent);
```
Please remove unused parameters.  This applies to all three of these.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -701,6 +719,71 @@ extern "C" HRESULT ElevationExecuteExePackage(
+{
+    UNREFERENCED_PARAMETER(hwndParent);
+    HRESULT hr = S_OK;
+    BURN_ELEVATION_MSI_MESSAGE_CONTEXT context = {};
```
This context isn't actually used so we should just pass in NULL for the context.  This applies to all three of these.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -701,6 +719,71 @@ extern "C" HRESULT ElevationExecuteExePackage(
+    context.pvContext = pvContext;
+
+    hr = PipeSendMessage(hPipe, BURN_ELEVATION_TRANSACTION_BEGIN, NULL, 0, NULL, &context, &dwResult);
+    ExitOnFailure(hr, "Failed to send BURN_ELEVATION_MESSAGE_TYPE_EXECUTE_MSI_PACKAGE message to per-machine process.");
```
The message should have BURN_ELEVATION_TRANSACTION_BEGIN, not BURN_ELEVATION_MESSAGE_TYPE_EXECUTE_MSI_PACKAGE.  Note that all three methods have the same mistake.

rseanhall commented on src/burn/engine/elevation.cpp:
```diff
@@ -1578,6 +1673,53 @@ static HRESULT ProcessResult(
+
+    pContext->hMsiTrns = NULL;
+    pContext->hMsiTrnsEvent = NULL;
+    uResult = MsiBeginTransaction(L"WiX", 0, &pContext->hMsiTrns, &pContext->hMsiTrnsEvent);
```
If we eventually need to store these values, we'll have to store it somewhere else (see comment above).

rseanhall commented on src/tools/wix/Bind/Bundles/OrderPackagesAndRollbackBoundariesCommand.cs:
```diff
@@ -45,10 +47,20 @@ public void Execute()
                             facade.Package.RollbackBoundary = previousRollbackBoundary.ChainPackageId;
                             previousRollbackBoundary = null;
+
+                            boundaryHadX86Package = (facade.Package.x64 == YesNoType.Yes);
```
I'm having trouble wrapping my head why this isn't `== YesNoType.No`.

rseanhall reviewed (commented):
Thanks, I'm going to go ahead and merge this in.  I've left several comments on things I'd like you to fix in your next pull request.  In the future, please add a history entry.  Also, there's a couple of additional things that I think need to be added:

1. The BA needs to be informed of these new actions.  I'm thinking a new message for BeginTransaction, CommitTransaction, and RollbackTransaction.  Also, it would probably help for the ExecutePackageBegin message to include whether it's inside of a transaction.

2. Rob mentioned somewhere that people don't really use `RollbackBoundary` elements and would like to see the `Transaction` attribute on the `Chain` element.  I'm thinking that it should be in both places - setting it on the `Chain` element would change the default for all `RollbackBoundary` elements as well as the automatically created ones.
